packages/protocol/scripts/truffle/make-release.ts

```diff
@@ -0,0 +1,244 @@
+  dependencies: Map<string, string[]>
+  constructor(libraries: any) {
+    this.dependencies = new Map()
+    Object.keys(libraries).forEach((lib: string) => {
+      libraries[lib].forEach((contract: string) => {
+        if (this.dependencies.has(contract)) {
+          this.dependencies.get(contract).push(lib)
+        } else {
+          this.dependencies.set(contract, [lib])
+        }
+      })
+    })
+  }
+
+  public get = (contract: string): string[] => {
+    return this.dependencies.has(contract) ? this.dependencies.get(contract) : []
+  }
+}
+
+class ContractAddresses {
+  static async create(contracts: string[], registry: any) {
+    const addresses = new Map()
+    await Promise.all(
+      contracts.map(async (contract: string) => {
+        const registeredAddress = await registry.getAddressForString(contract)
+        if (!eqAddress(registeredAddress, NULL_ADDRESS)) {
+          addresses.set(contract, registeredAddress)
+        }
+      })
+    )
+    return new ContractAddresses(addresses)
+  }
+
+  constructor(public addresses: Map<string, Address>) {}
+
+  public get = (contract: string): Address => {
+    if (this.addresses.has(contract)) {
+      return this.addresses.get(contract)
+    } else {
+      throw new Error(`Unable to find address for ${contract}`)
+    }
+  }
+
+  public set = (contract: string, address: Address) => {
+    this.addresses.set(contract, address)
+  }
+}
+
+const REGISTRY_ADDRESS = '0x000000000000000000000000000000000000ce10'
+
+module.exports = async (callback: (error?: any) => number) => {
+  try {
+    const argv = require('minimist')(process.argv.slice(2), {
+      string: ['report', 'network', 'proposal', 'libraries', 'initialize_data', 'build_directory'],
+      boolean: ['dry_run'],
+    })
+    const report: ASTDetailedVersionedReport = readJsonSync(argv.report).report
+    const initializationData = readJsonSync(argv.initialize_data)
+    const dependencies = new ContractDependencies(linkedLibraries)
+    const contracts = readdirSync(join(argv.build_directory, 'contracts')).map((x) =>
+      basename(x, '.json')
+    )
+    const registry = await artifacts.require('Registry').at(REGISTRY_ADDRESS)
+    const addresses = await ContractAddresses.create(contracts, registry)
+    const released: Set<string> = new Set([])
+    const proposal = []
+    // Release 1 will deploy all libraries with proxies so that they're more easily
+    // upgradable. All contracts that link libraries should be upgraded to instead link the proxied
+    // library.
+    // To ensure this actually happens, we check that all contracts that link libraries are marked
+    // as needing to be redeployed.
+    // TODO(asa): Remove this check after release 1.
+    const linksLibrariesWithoutChanges = contracts.map(
+      (c) =>
+        dependencies.get(c).length &&
+        !Object.keys(report.contracts).includes(c) &&
+        !c.endsWith('Test')
+    )
+    if (linksLibrariesWithoutChanges.some((b) => b)) {
```

Comment:
Optional Nit: I personally had trouble following the code, what do you think about this one

```typescript

function ensureAllContractsWithLibrariesAreUpgraded() {
  let anyContractViolates = false
  contracts.map((contract) => {
    const hasDependency = dependencies.get(contract).length > 0
    const specifiedToUpgrade = report.contracts[contract]
    const isTest = contract.endsWith("Test")

    if (hasDependency && !specifiedToUpgrade && !isTest) {
      console.log(
        `${contract} links ${dependencies.get(contract)} and needs to be upgraded to link proxied libraries.`
      )
      anyContractViolates = true
    }
  });

  if (anyContractViolates) {
    throw new Error('All contracts linking libraries should be upgraded in release 1')
  }
}


```